ma_gym/envs/lumberjacks/lumberjacks.py

```diff
@@ -0,0 +1,351 @@
+                write_cell_text(img, text=str(agent_strength), pos=agent_pos,
+                                cell_size=cell_size, fill='white', margin=0.4)
+
+        img = np.asarray(img)
+        if mode == 'rgb_array':
+            return img
+        elif mode == 'human':
+            from gym.envs.classic_control import rendering
+            if self._viewer is None:
+                self._viewer = rendering.SimpleImageViewer()
+            self._viewer.imshow(img)
+            return self._viewer.isopen
+
+    def _view_generator(self, mask: Tuple[slice, slice]) -> Tuple[Coordinates, int, int]:
+        """Yields position, number of agent and tree strength for all cells defined by `mask`.
+
+        Args:
+            mask: tuple of slices in extended coordinates.
+        """
+        agent_iter = np.ndenumerate(np.sum(self._agent_map[mask], axis=2))
+        tree_iter = np.nditer(self._tree_map[mask])
+        for (pos, n_a), n_t in zip(agent_iter, tree_iter):
+            yield pos, n_a, n_t
+
+    def get_agent_obs(self) -> List[List[float]]:
+        """Returns list of observations for each agent."""
+        obs = np.zeros((self.n_agents, self._obs_len))
+        for i, (agent_id, agent) in enumerate(self._agent_generator()):
+            rel_pos = self._to_relative_coordinates(agent.pos)
+            obs[i, 0] = agent_id / self.n_agents  # Agent ID
+            obs[i, 1] = rel_pos[0] / (self._grid_shape[0] - 1)  # Coordinate
+            obs[i, 2] = rel_pos[1] / (self._grid_shape[1] - 1)  # Coordinate
+            obs[i, 3] = self._step_count / self._max_steps  # Steps
+
+            for j, (_, agent_strength, tree_strength) in zip(
+                    itertools.count(start=4, step=2),
+                    self._agent_view_generator(agent.pos, self._agent_view)):
+                obs[i, j] = agent_strength / self.n_agents
+                obs[i, j+1] = tree_strength / self.n_agents
+
+        # Convert it from numpy array
+        obs = obs.tolist()
+
+        if self.full_observable:
+            obs = [feature for agent_obs in obs for feature in agent_obs]
+            obs = [obs] * self.n_agents
+
+        return obs
+
+    def _agent_generator(self) -> Tuple[int, Agent]:
+        """Yields agent_id and agent for all agents in environment."""
+        for agent_id, agent in enumerate(self._agents):
+            yield agent_id, agent
+
+    def _agent_view_generator(self, pos: Coordinates, view_range: Tuple[int, int]):
+        """Yields position, number of agent and tree strength for cells in distance of `view_range` from `pos`.  """
+        mask = (
+            slice(pos[0] - view_range[0], pos[0] + view_range[0] + 1),
+            slice(pos[1] - view_range[1], pos[1] + view_range[1] + 1),
+        )
+        yield from self._view_generator(mask)
+
+    def step(self, agents_action: List[int]):
+        assert len(agents_action) == self.n_agents
+
```

Comment:
Also, assert that every action is a valid action.

Like user may send [-10,10,1,2]

Reply:
I am not sure if having asserts in `step()` method is a good idea at all. As this method is called every time, the assert check would significantly slow down the environment. Moreover, the environment would raise an exception in any case on line 309.

Reply:
👍  I have added a comment on line 309